vhost-user-backend/src/handler.rs

```diff
@@ -615,6 +623,53 @@ where
 
         Ok(())
     }
+
+    fn postcopy_advice(&mut self) -> VhostUserResult<File> {
+        let mut uffd_builder = UffdBuilder::new();
+
+        let uffd = uffd_builder
+            .close_on_exec(true)
+            .non_blocking(true)
+            .user_mode_only(false)
+            .create()
+            .map_err(|e| {
+                VhostUserError::ReqHandlerError(io::Error::new(io::ErrorKind::Other, e))
+            })?;
+
+        // SAFETY:
+        // We know that uffd is a valid fd.
+        let uffd_file = unsafe { File::from_raw_fd(uffd.as_raw_fd()) };
+
+        self.uffd = Some(uffd);
+
+        Ok(uffd_file)
+    }
+
+    fn postcopy_listen(&mut self) -> VhostUserResult<()> {
+        let Some(ref uffd) = self.uffd else {
+            return Err(VhostUserError::ReqHandlerError(io::Error::new(
+                io::ErrorKind::Other,
+                "No registered UFFD handler",
+            )));
+        };
+
+        for mapping in self.mappings.iter() {
+            uffd.register(
+                mapping.local_addr as *mut libc::c_void,
```

Comment:
Hey, sorry about missing this completely somehow.. So yes, I had a look at code again.. With Xen's Grant memory mapping (mapping on the fly once the buffers are available), this would lead to:

https://github.com/rust-vmm/vm-memory/blob/main/src/mmap_xen.rs#L890

which will return `null_mut()`.

The entire region isn't mapped in advance and we get to that only when we get the final request.

Reply:
Hmm.. so how would vhost-backend even be configured with Xen mappings? It is just in this case mappings are created here https://github.com/rust-vmm/vhost/blob/321085b2ea7bd818942be7954f3daaba1fa34293/vhost-user-backend/src/handler.rs#L282. The backend implementation in this repo does not use Xen mappings, so there is should be no way for the pointer to be `null`. Am I missing smth?

Reply:
> The backend implementation in this repo does not use Xen mappings, so there is should be no way for the pointer to be `null`. Am I missing smth?

The `xen` feature of the crate allows enabling the Xen mapping. The vhost-device daemons (or virtiofsd) use that for Xen support. With Xen support no virtual address is known until a slice is attempted to be accessed (at which point it will be mapped only for the needed access duration).

So I think this would probably need to be integrated with vm-memory (which handles the mapping and unmapping) to some degree. Maybe the existing dirty page tracking may help there...

Reply:
So do I understand correctly that when Xen is used with Postcopy feature, the registration for uffd should happen right after memory is mapped. And unregistered after operation is finished before unmapping? This seems like a lot of movement and overall I am not sure if anyone will use this specific combination.

What if I restrict the postcopy support to be only availabe when Xen feature is not used?

Reply:
> So do I understand correctly that when `xen` is used with `VHOST_USER_PROTOCOL_F_PAGEFAULT` feature, the registration for uffd should happen right after memory is mapped. And unregistered after operation is finished before unmapping? This seems like a lot of movement and overall I am not sure if anyone will use this specific combination.

I am not sure whether that is what should be happening, I only pointed out that that is the only period over which the mapping guaranteed to be known.

> What if I restrict the postcopy support to be only availabe when `xen` feature is not used?

IMHO it is not so much a Xen question, but more of an architecture question. Xen (and I was told pKVM too) need "on-demand" mapping. Hence, we deprecated other places that allowed direct pointer access without mapping guards in place. But I will have to leave it to someone more familiar with the technology to comment on how this would be best modeled in an hypervisor-agnostic way.

Reply:
@Ablu @vireshk I don't know Xen at all, so what do you think about the current implementation (postcopy supported only when `xen` feature is not enabled) ?